sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -620,17 +621,28 @@ class HDFSFileCatalog(
     if (paths.length >= sqlContext.conf.parallelPartitionDiscoveryThreshold) {
       HadoopFsRelation.listLeafFilesInParallel(paths, hadoopConf, sqlContext.sparkContext)
     } else {
-      val statuses = paths.flatMap { path =>
+      val statuses: Seq[FileStatus] = paths.flatMap { path =>
         val fs = path.getFileSystem(hadoopConf)
         logInfo(s"Listing $path on driver")
         // Dummy jobconf to get to the pathFilter defined in configuration
-        val jobConf = new JobConf(hadoopConf, this.getClass())
+        val jobConf = new JobConf(hadoopConf, this.getClass)
         val pathFilter = FileInputFormat.getInputPathFilter(jobConf)
-        if (pathFilter != null) {
+        val statuses = if (pathFilter != null) {
           Try(fs.listStatus(path, pathFilter)).getOrElse(Array.empty)
         } else {
           Try(fs.listStatus(path)).getOrElse(Array.empty)
         }
+
+        fs match {
+          case _: S3FileSystem =>
```

Comment:
Actually, do we need to have this pattern matching? Looks like S3 filesystems just use the default implementation of getFileBlockLocations, which returns `localhost:50010`?